lib/functions.php

```diff
@@ -752,9 +752,20 @@ function coroutine(callable $func) {
  * Upon resolution the Generator return value is used to succeed the promised result. If an
  * error occurs during coroutine resolution the returned promise fails.
  *
- * @param \Generator $generator The generator to resolve as a coroutine
+ * @param \Generator|callable $generator The generator to resolve as a coroutine
+ * @return \Amp\Promise
  */
-function resolve(\Generator $generator) {
+function resolve($generator) {
+    if (is_callable($generator)) {
+        $generator = call_user_func($generator);
+
+        if (!$generator instanceof \Generator) {
+            throw new \LogicException('Callable passed to resolve() did not return a Generator');
+        }
+    } else if (!$generator instanceof \Generator) {
+        throw new \LogicException('Co-routine to resolve must be callable or Generator');
```

Comment:
I agree that should be an `InvalidArgumentException` on L766 (as if the spl extension model makes any sense at all :-P).

WRT coroutine vs co-routine I'll go with the project standard, from an English perspective Co-routine is probably more correct but consistency is more important.